File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/anytype/Dereference.java
/*
 *  Copyright 2024 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package hu.bme.mit.theta.core.type.anytype;

import hu.bme.mit.theta.common.Utils;
import hu.bme.mit.theta.core.model.Valuation;
import hu.bme.mit.theta.core.type.Expr;
import hu.bme.mit.theta.core.type.LitExpr;
import hu.bme.mit.theta.core.type.Type;
import hu.bme.mit.theta.core.type.inttype.IntType;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

import static com.google.common.base.Preconditions.checkState;
import static hu.bme.mit.theta.core.utils.TypeUtils.cast;

public final class Dereference<A extends Type, T extends Type> implements Expr<T> {

    private static final String OPERATOR_LABEL = "deref";
    private static int counter = 0;
    private final Expr<A> array;
    private final Expr<A> offset;
    private final T type;

    private final Optional<Expr<IntType>> uniquenessIdx;

    private Dereference(Expr<A> array, Expr<A> offset, T type) {
        this.array = array;
        this.offset = offset;
        this.type = type;
        uniquenessIdx = Optional.empty();
    }


    private Dereference(Expr<A> array, Expr<A> offset, Expr<IntType> uniqueness, T type) {
        this.array = array;
        this.offset = offset;
        this.type = type;
        this.uniquenessIdx = Optional.of(uniqueness);
    }

    public Expr<A> getArray() {
        return array;
    }

    public Expr<A> getOffset() {
        return offset;
    }


    public static <A extends Type, T extends Type> Dereference<A, T> of(Expr<A> array, Expr<A> offset, T type) {
        return new Dereference<>(array, offset, type);
    }

    private static <A extends Type, T extends Type> Dereference<A, T> of(Expr<A> array, Expr<A> offset, Expr<IntType> uniqueness, T type) {
        return new Dereference<>(array, offset, uniqueness, type);
    }

    public Dereference<A, T> withUniquenessExpr(Expr<IntType> expr) {
        return Dereference.of(array, offset, expr, type); // TODO: this kills the stuck check
    }

    @Override
    public int getArity() {
        return 3;
    }

    @Override
    public T getType() {
        return type;
    }

    @Override
    public LitExpr<T> eval(Valuation val) {
        throw new IllegalStateException(
                "Reference/Dereference expressions are not meant to be evaluated!");
    }

    @Override
    public List<? extends Expr<?>> getOps() {
        return uniquenessIdx.isPresent() ? List.of(array, offset, uniquenessIdx.get()) : List.of(array, offset);
    }

    @Override
    public Expr<T> withOps(List<? extends Expr<?>> ops) {
        checkState(ops.size() == 3 || ops.size() == 2);
        @SuppressWarnings("unchecked") final T ptrType = (T) ops.get(0).getType();
        if (ops.size() == 3) {
            return Dereference.of(cast(ops.get(0), ptrType), cast(ops.get(1), ptrType), (Expr<IntType>) ops.get(2), type);
        } else {
            return Dereference.of(cast(ops.get(0), ptrType), cast(ops.get(1), ptrType), type);
        }
    }

    @Override
    public int hashCode() {
        return Objects.hash(array, offset, uniquenessIdx, type);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Dereference<?, ?> that) {
            return Objects.equals(this.array, that.array) &&
                    Objects.equals(this.offset, that.offset) &&
                    Objects.equals(this.uniquenessIdx, that.uniquenessIdx) &&
                    Objects.equals(this.type, that.type);
        }
        return false;
    }

    @Override
    public String toString() {
        var base = Utils.lispStringBuilder(OPERATOR_LABEL).add(getArray()).add(getOffset());
        uniquenessIdx.ifPresent(base::add);
        return base.add(type).toString();
    }

    public Optional<Expr<IntType>> getUniquenessIdx() {
        return uniquenessIdx;
    }
}

File: subprojects/common/core/src/main/kotlin/hu/bme/mit/theta/core/type/anytype/Dereference.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.core.type.anytype

import hu.bme.mit.theta.common.Utils
import hu.bme.mit.theta.core.model.Valuation
import hu.bme.mit.theta.core.type.Expr
import hu.bme.mit.theta.core.type.LitExpr
import hu.bme.mit.theta.core.type.Type
import hu.bme.mit.theta.core.type.inttype.IntType
import kotlinx.serialization.*
import kotlinx.serialization.builtins.nullable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.*

/**
 * Represents a dereference expression for array access.
 *
 * @param A The array type
 * @param O The offset type
 * @param T The element type
 * @property array The array expression
 * @property offset The offset expression
 * @property type The type of the dereferenced element
 * @property uniquenessIdx Optional uniqueness index for SMT encoding
 */
@Serializable(with = Dereference.Serializer::class)
@SerialName("Dereference")
data class Dereference<A : Type, O : Type, T : Type>(
  val array: Expr<A>,
  val offset: Expr<O>,
  override val type: T,
  val uniquenessIdx: Expr<IntType>? = null,
) : Expr<T> {

  companion object {

    private const val OPERATOR_LABEL = "deref"

    @JvmStatic
    fun <A : Type, O : Type, T : Type> of(
      array: Expr<A>,
      offset: Expr<O>,
      type: T,
    ): Dereference<A, O, T> = Dereference(array, offset, type)

    @JvmStatic
    private fun <A : Type, O : Type, T : Type> of(
      array: Expr<A>,
      offset: Expr<O>,
      uniqueness: Expr<IntType>,
      type: T,
    ): Dereference<A, O, T> = Dereference(array, offset, type, uniqueness)
  }

  fun withUniquenessExpr(expr: Expr<IntType>): Dereference<A, O, T> = of(array, offset, expr, type)

  override val arity: Int = 3

  override val ops: List<Expr<*>> =
    if (uniquenessIdx != null) listOf(array, offset, uniquenessIdx) else listOf(array, offset)

  override fun eval(`val`: Valuation): LitExpr<T> =
    throw IllegalStateException("Reference/Dereference expressions are not meant to be evaluated!")

  @Suppress("UNCHECKED_CAST")
  override fun withOps(ops: List<Expr<*>>): Expr<T> {
    require(ops.size == 2 || ops.size == 3) { "Dereference must have 2 or 3 operands" }
    return when (ops.size) {
      2 -> of(ops[0] as Expr<A>, ops[1] as Expr<O>, type)
      else -> of(ops[0] as Expr<A>, ops[1] as Expr<O>, ops[2] as Expr<IntType>, type)
    }
  }

  override fun toString(): String =
    Utils.lispStringBuilder(OPERATOR_LABEL).body().addAll(ops).add(type).toString()

  object Serializer : KSerializer<Dereference<out Type, out Type, out Type>> {

    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("Dereference") {
      element<Expr<out Type>>("array")
      element<Expr<out Type>>("offset")
      element<Type>("type")
      element<Expr<IntType>?>("uniquenessIdx")
    }

    override fun serialize(encoder: Encoder, value: Dereference<out Type, out Type, out Type>) =
      encoder.encodeStructure(descriptor) {
        encodeSerializableElement(descriptor, 0, PolymorphicSerializer(Expr::class), value.array)
        encodeSerializableElement(descriptor, 1, PolymorphicSerializer(Expr::class), value.offset)
        encodeSerializableElement(descriptor, 2, PolymorphicSerializer(Type::class), value.type)
        encodeSerializableElement(descriptor, 3, PolymorphicSerializer(Expr::class).nullable, value.uniquenessIdx)
      }

    override fun deserialize(decoder: Decoder): Dereference<out Type, out Type, out Type> =
      decoder.decodeStructure(descriptor) {
        var array: Expr<out Type>? = null
        var offset: Expr<out Type>? = null
        var type: Type? = null
        var uniquenessIdx: Expr<IntType>? = null
        while (true) {
          when (val i = decodeElementIndex(descriptor)) {
            0 -> array = decodeSerializableElement(descriptor, 0, PolymorphicSerializer(Expr::class))
            1 -> offset = decodeSerializableElement(descriptor, 1, PolymorphicSerializer(Expr::class))
            2 -> type = decodeSerializableElement(descriptor, 2, PolymorphicSerializer(Type::class))
            3 -> uniquenessIdx =
              decodeSerializableElement(descriptor, 3, PolymorphicSerializer(Expr::class).nullable) as Expr<IntType>?

            CompositeDecoder.DECODE_DONE -> break
            else -> throw SerializationException("Unknown index $i")
          }
        }
        Dereference(
          array ?: throw SerializationException("Missing array "),
          offset ?: throw SerializationException("Missing offset "),
          type ?: throw SerializationException("Missing type "),
          uniquenessIdx
        )
      }
  }
}
